### PR TITLE
Tidy up new React + API explorer for mobile (fixes #3607)

### LINKED ISSUE #3607: Tidy up new React + API explorer
Tracking issue for the follow-up work to #3012, #3385. We now need to tidy up the last few pieces before the 1.11 release.

__Help is very welcome so we can release this on time__. This is an idiomatic React + Redux codebase, so a great way to get up to speed with this stack if beginners want to get involved.

### Remaining issues

- [ ] On mobile screen sizes, the last items in the menu require page scrolling as well as menu scrolling.
- [x] On mobile screen sizes, top part of explorer behind address bar.
- ~[ ] On mobile, page overlay doesn't show.~ (desired)
- [x] On touch screens, tweak hover/active styles to not show as often.
- [x] On touch screens, explorer dismissal on touch feels a bit too fast?
- [x] IE11 - Menu items are slightly wider than their container.
- [x] IE11 - Menu items load stacked onto one another.
- ~[ ] Windows – Scrollbars should be hidden during transitions.~ (no, same across all browsers)
- [x] On mobile, Close explorer link does not work, clicking it just brings you back to the first explorer 'home page'
- [x] On mobile, Click on “Pages” the page scrolls down, so I scroll up, then click on “Close explorer”, which closes the explorer but reopens it immediately, while scrolling down again.
- ~[ ] On mobile, on first load of a menu level, focus ends on "Close explorer" button.~ (desired)
- [x] On mobile, explorer shows a shadow even though there is no layering on the page.
- [x] Performance review (CPU and memory, of both the Redux and React hierarchies)

### Features

- [ ] Text truncation of the page title after 60-70 characters;
- ~[ ] Finishing adding the setting to customise the ordering / API behavior – I've added this as a global variable initialised in the templates, but it's not hooked into the Django settings yet (I don't know how to do that)~

## Backlog

- [ ] Code reviews

### Documentation

- [ ] React/Redux code styleguide (action types with keyMirror, FSA or not, file structure)
- [ ] Integrate React components in UI styleguide

### Features

- [ ] Better translation with jsi18n (https://docs.google.com/document/d/1WImkcI1ORB5nj3FeNuVnkPIcjJTIKoMIhCV-Hqj2IEY)
- ~[ ] Better date format client-side (`moment-django` FTW!)~

### Tooling

- [ ] Add front-end test coverage to Codecov
- [x] Update to Webpack 2
- [x] Update to latest Jest
- [x] Update to latest React
- [ ] npm packaging of Wagtail front-end  https://www.npmjs.com/package/wagtail
- [ ] Reach 100% unit test coverage

### Related work to address later

- [ ] @kaedroho's page chooser refactor #3038
- [ ] #1160 
- [ ] #3106 
- [ ] #3442

## Cross-browser test suite (to copy/paste when doing a test run)

- [ ] - Browser	Device/OS	Version(s)
- [ ] - Mobile Safari	iOS Phone	10
- [ ] - Mobile Safari	iOS Phone	9
- [ ] - Mobile Safari	iOS Tablet	10
- [ ] - Mobile Safari	iOS Tablet	9
- [ ] - Chrome	Android	56
- [ ] - Chrome	Android	55
- [ ] - IE	Desktop	11
- [ ] - Chrome	Desktop	56
- [ ] - Chrome	Desktop	55
- [ ] - MS Edge	Desktop	15
- [ ] - MS Edge	Desktop	14
- [ ] - Firefox	Desktop	52
- [ ] - Firefox ESR	Desktop	45.8
- [ ] - Safari	macOS	9
- [ ] - Safari	macOS	8 #

### PR DESCRIPTION
**Pull request by thibaudcolas**
Most of the work from #3607 to make the explorer work better (or at all) on mobile, and also faster everywhere.

Edit: not WIP anymore.

### CODE REVIEW

File: client/src/components/Explorer/ExplorerPanel.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ import { STRINGS, MAX_EXPLORER_PAGES } from '../../config/wagtailConfig';
 
 import Button from '../Button/Button';
 import LoadingSpinner from '../LoadingSpinner/LoadingSpinner';
-import Transition, { PUSH, POP, FADE } from '../Transition/Transition';
```
Comment (jjanssen) [inline client/src/components/Explorer/ExplorerPanel.js]: If we are not using `FADE` at the moment, we can clean this up from `Transition.js` and `Transition.scss` as well for now.

File: gulpfile.js/tasks/styles.js [hunk 2]
```diff
@@ -33,7 +33,7 @@ gulp.task('styles:sass', function () {
             outputStyle: 'expanded'
         }).on('error', sass.logError))
         .pipe(autoprefixer({
-            browsers: ['last 3 versions', 'not ie <= 8'],
+            browsers: ['last 3 versions', 'ie 11'],
```
Comment (jjanssen) [inline gulpfile.js/tasks/styles.js]: 🎉

File: client/src/components/Explorer/reducers/nodes.js [hunk 3]
```diff
@@ -15,26 +15,30 @@ const defaultPageState = {
...
 export default function nodes(prevState = defaultState, { type, payload }) {
-  const state = Object.assign({}, prevState);
+  let state = prevState;
 
   switch (type) {
   case 'OPEN_EXPLORER':
+    state = Object.assign({}, prevState);
     state[payload.id] = Object.assign({}, defaultPageState, state[payload.id]);
-    break;
+    return state;
```
Comment (jjanssen) [inline client/src/components/Explorer/reducers/nodes.js]: I like this, looks more consolidated.

File: client/src/components/Explorer/ExplorerPanel.js (hunk 1 above)
Comment (thibaudcolas) [inline client/src/components/Explorer/ExplorerPanel.js]: 👍

**Comment by thibaudcolas** [comment]
@jjanssen I think I'm done. There are only two things left on #3607 that I'd like to take care of before the next release, but they will be in separate PRs.

Looking over all of my changes, it's a bit messy – I cleaned up as much as I could to keep the reducers small and fast, and also refactored some of the styles (on "mobile" screen sizes mainly).

**Comment by thibaudcolas** [comment]
Yes! Merged 🙂 . There is one last PR at #3645, also detailing how much testing I've done so far. I'm quite confident it works (haha), but any further help is very much appreciated.

**Pull request merged**